docs/GettingStartedDocs/Contributors/SignUsingEngines.md

```diff
@@ -0,0 +1,114 @@
+# Signing a Build With OpenSSL Engines
+
+OpenSSL allows extensions called "engines" to perform the generation of keys, signing and encryption.
+Engines are useful. They can be used to allow hardware offload of encryption duties, or improved 
+encapsulation of secrets. Use of engines is required in some build systems as a way to avoid the
+storage of keys in plain text .pem files.
+
+oesign has been equipped with the ability to specify engines in the enclave signing process. In order
+to use the engine, the command line syntax has been extended.
+```
+oesign sign --enclave-image ENCLAVE_IMAGE --config-file CONFIG_FILE --engine engine-NAME -load-path engine-LOAD-PATH-key KEY-ID
+
+```
+
+## Engines 
+
+The concept of an engine was added to OpenSSL starting with 0.9.6. There are built in engine implementations
+implementations for
+- Microsoft CryptoAPI
+- VIA Padlock
+- nCipher CHIL
+and a api for controlling dynamically bound engines which are bound from shared libraries at runtime via either program logic
+or configuration files. 
+
+The openssl commnd line utility can show available engines via the "engine" command.
+```
+# openssl engine
+(rdrand) Intel RDRAND engine
+(dynamic) Dynamic engine loading support
+
+```
+Of particular interest is the "dynamic" engine, which allow dynamic plugins to perform encryption duties. 
+These objects allow a plugin mechanism for implementations of cryptographic algorithms and can be statically linked in the
+OpenSSL library at compile-time or dynamically loaded at run-time in and out of the running application 
+with low overhead from external binary shared objects implementing the engine API using special built-in engine
+called “dynamic”. 
+
+Dynamic engines are particularly interesting due to the flexibility they provide: 
+
+- They allow sensitive encryption data structures, such as private key bits, or data to be signed, to be kept out of publicly viewable memory, 
+  so reduce the risk of leaked secrets in case of a security breach. The actual secrets may be kept in enclave memory or another dedicated host service.
+
+- They allow sensitive processes and algorithms to be divided from the overall project and hidden from other organizations.
+
+- They reduce the memory impact of OpenSSL, by avoiding statically linking support for unneeded hardware or features at
+  compile-time in favor of system configuration or automatically probing for supported devices at run-time and dynamically
+  loading only the required cryptographic modules.
+
+- They allow the developer or system administrator to replace compiled in OpenSSL functionality in case of bugs, 
+  known vulnerabilities or sub-optimal performances with more appropriate alternative implementations while maintaining compatibility
+  with existing applications.
+
+- They provide an alternative in case of issues with the OpenSSL core development team or distro vendor decision-making process, decoupling
```

Comment:
What's the relationship between openssl and mbedtls?  Do we really need to have the OE SDK depend on *both* for different purposes, instead of only one and making that one be replacable by the other?

Reply:
There is no engine facility in mbedtls, so that reduces the question to do we need mbedtls. 

That seems an interesting question but outside the scope of this PR, no?

Reply:
I agree the question is technically on the design doc for this and not the code PR per se, but I didn't see a design doc PR to comment on.  If there is one, can you point me to it?  (I looked in https://github.com/openenclave/openenclave/tree/master/docs/DesignDocs but nothing jumped out at me as being it.)

Reply:
I don't understand what you are asking here.
Are you suggesting that oesign get completely rewritten to use mbedtls?

Reply:
host side uses OpenSSL by default (linux) and enclave uses mbedtls by default. OEsign is a host application and is already coded against OpenSSL. Nothing has changed in this pull request around this. Therefore if you have this question it seems like it should be in a separate issue?

Reply:
I am ok with Paul's answer (host side uses openssl and enclave uses mbedtls) for now, but the meta point is that we need something in the design docs that actually says/explains that, and I don't see it.  I am fine with a separate PR for that.